### PR TITLE
Remove compile time dependency on Microsoft.WindowsAzure.ServiceRuntime

### PR DESCRIPTION
**Pull request by sergeybykov**
This is a bit crazy by seems to work.

The compile time dependency on `Microsoft.WindowsAzure.ServiceRuntime.dll` that lacks an official NuGet package and its versioning are a major pain in building Azure-based solutions. The APIs we are using from `Microsoft.WindowsAzure.ServiceRuntime` have been stable for years. So why not use it via Reflection and not worry about which version it is when they all are really the same?

In a worker role, `Microsoft.WindowsAzure.` is normally already loaded by Azure role runtime, so we just need to find it in memory.

In a web role, it may or may not be already loaded. This is where I fall back to using `Assembly.LoadWithPartialName`, which is deprecated. Not sure if there is a better way to load an known assembly of any version available.

### CODE REVIEW

File: src/OrleansAzureUtils/AzureSilo.cs [hunk 1]
```diff
@@ -150,29 +130,29 @@ public bool Start(string deploymentId, RoleInstance myRoleInstance, ClusterConfi
 
             myEntry = new SiloInstanceTableEntry
             {
-                DeploymentId = deploymentId,
-                Address = myEndpoint.IPEndpoint.Address.ToString(),
-                Port = myEndpoint.IPEndpoint.Port.ToString(CultureInfo.InvariantCulture),
+                DeploymentId = serviceRuntimeWrapper.DeploymentId,
+                Address = myEndpoint.Address.ToString(),
```
Comment (gabikliot) [inline src/OrleansAzureUtils/AzureSilo.cs]: because of line like this it is important NOT to use var in "var myEndpoint = serviceRuntimeWrapper.GetEndpoint(SiloEndpointConfigurationKeyName);"
How am I supposed to guess that you changed  serviceRuntimeWrapper.GetEndpoint to return IPEndpoint while it used to be RoleInstanceEndpoint ?

It is definitely not true that by looking at `var myEndpoint = serviceRuntimeWrapper.GetEndpoint(SiloEndpointConfigurationKeyName);` it obvious to deduce the type of myEndpoint .

Bottom line: don't use var there.

Comment (sergeybykov) [inline src/OrleansAzureUtils/AzureSilo.cs]: Alternatively, I could rename `GetEndpoint()` to `GetIPEndpoint()` to be more explicit.

Comment (sergeybykov) [inline src/OrleansAzureUtils/AzureSilo.cs]: Done.

**Comment by sergeybykov** [comment]
Inspired by @veikkoeeva I replaced some of the ugliest Reflection code with dynamic. Doesn't seem to work for statics though.

**Comment by veikkoeeva** [comment]
@sergeybykov Would [this](http://blogs.msdn.com/b/davidebb/archive/2009/10/23/using-c-dynamic-to-call-static-members.aspx) help? I think you need (without delving into the code) a ``DynamicObject`` helper.

**Comment by sergeybykov** [comment]
@veikkoeeva Thanks for the pointer. An interesting hack. I don't think it's worth it for this case though as it won't save me any code really. But good to know for the future.

**Comment by jthelin** [comment]
LGTM.
